File: hyperscalp-ai-main/research/backtester/strategies/liquidation_fade.py
"""Python port of strategies/liquidationFade.ts."""
from dataclasses import dataclass
from typing import Optional
import pandas as pd

from ..features import atr


@dataclass
class Signal:
    strategy: str; direction: str
    entry: float; stop: float; target: float
    expected_holding_minutes: int
    preferred_entry_mode: str
    reason: str


def liquidation_fade_signal(bars1m: pd.DataFrame, trades: pd.DataFrame, liquidations: pd.DataFrame, coin: str, cfg: dict) -> Optional[Signal]:
    if len(bars1m) < 30 or liquidations.empty:
        return None
    end_ts = bars1m.index[-1]
    window = liquidations[(liquidations["coin"] == coin) & (liquidations["ts"] >= end_ts - pd.Timedelta(seconds=5)) & (liquidations["ts"] <= end_ts)]
    if window.empty:
        return None
    long_usd = float(window.loc[window["side"] == "long", "usd"].sum())
    short_usd = float(window.loc[window["side"] == "short", "usd"].sum())
    net = abs(long_usd - short_usd)
    if net < cfg.get("min_liquidation_usd_5s", 250_000):
        return None

    direction = "long" if long_usd > short_usd else "short"

    lb = cfg.get("sweep_lookback_min", 5)
    recent = bars1m.iloc[-(lb + 1):-1]
    if recent.empty:
        return None
    prev_high = float(recent["high"].max())
    prev_low = float(recent["low"].min())
    last = bars1m.iloc[-1]
    swept = (last["low"] < prev_low) if direction == "long" else (last["high"] > prev_high)
    if not swept:
        return None

    if cfg.get("require_aggressor_flip", True) and not trades.empty:
        # Last 10s aggressor imbalance
        sub = trades[trades["ts"] >= end_ts - pd.Timedelta(seconds=10)]
        if not sub.empty:
            buy = sub.loc[sub["is_buy"], "sz"].sum()
            sell = sub.loc[~sub["is_buy"], "sz"].sum()
            tot = buy + sell
            imb = (buy - sell) / tot if tot > 0 else 0
            if direction == "long" and imb < 0.10:
                return None
            if direction == "short" and imb > -0.10:
                return None

    price = float(last["close"])
    a1 = atr(bars1m, 14)
    stop = prev_low - a1 * 0.3 if direction == "long" else prev_high + a1 * 0.3
    r = abs(price - stop)
    target = price + r * cfg.get("rr_target", 1.5) if direction == "long" else price - r * cfg.get("rr_target", 1.5)

    return Signal(
        strategy="liquidation_fade", direction=direction, entry=price, stop=stop, target=target,
        expected_holding_minutes=cfg.get("max_holding_minutes", 8),
        preferred_entry_mode="ioc",
        reason=f"LIQ_FADE {direction.upper()} long_liq=${long_usd:.0f} short_liq=${short_usd:.0f}",
    )

File: hyperscalp-ai-main/base44/functions/botExecutor/strategies/liquidationFade.ts
/**
 * Strategy C — Liquidation Cascade Fade.
 *
 * Thesis: when forced liquidations cascade in one direction, the move
 * overshoots fundamentals because liquidation orders are price-insensitive
 * market orders. As soon as the cascade exhausts, the move snaps back.
 *
 * Trigger:
 *   - large net liquidation USD in a 5s window
 *   - 1m sweep of the session extreme in the same direction as the liquidations
 *   - aggressor flip in the opposite direction within the next few seconds (cascade is over)
 *   - take the opposite side
 *
 * Exit: 1.5R RR, no trailing — quick scalp. Time stop in 8 minutes regardless.
 */
import type { StrategyContext, StrategyResult } from './types.ts';
import { atr, sweepCheck, aggressorFlip, liquidationCascade } from '../lib/features.ts';

export function liquidationFade(ctx: StrategyContext): StrategyResult {
  const cfg = ctx.cfg;
  if (!cfg?.enabled) return { signal: null, reject: 'disabled' };
  if (ctx.c1m.length < 30) return { signal: null, reject: 'insufficient 1m candles' };

  const cascade = liquidationCascade(ctx.liquidations, ctx.coin, 5_000);
  if (Math.abs(cascade.longUsd - cascade.shortUsd) < cfg.min_liquidation_usd_5s) {
    return { signal: null, reject: `cascade $${Math.abs(cascade.longUsd - cascade.shortUsd).toFixed(0)} < min` };
  }
  // Direction: longs liquidating = price was dumping = fade by buying
  const direction: 'long' | 'short' = cascade.longUsd > cascade.shortUsd ? 'long' : 'short';

  const sweep = sweepCheck(ctx.c1m, cfg.sweep_lookback_min ?? 5);
  const sweptCorrectSide = direction === 'long' ? sweep.sweptLow : sweep.sweptHigh;
  if (!sweptCorrectSide) return { signal: null, reject: 'no sweep of session extreme on correct side' };

  if (cfg.require_aggressor_flip) {
    const flip = aggressorFlip(ctx.tape, 10_000, 60_000);
    const flipCorrect = direction === 'long' ? flip > 0.10 : flip < -0.10;
    if (!flipCorrect) return { signal: null, reject: `no aggressor flip (${flip.toFixed(2)})` };
  }

  const last = ctx.c1m[ctx.c1m.length - 1];
  const price = last.c;
  const a1 = atr(ctx.c1m, 14);
  // Stop just past the swept level
  const stop = direction === 'long' ? sweep.prevLow - a1 * 0.3 : sweep.prevHigh + a1 * 0.3;
  const r = Math.abs(price - stop);
  const target = direction === 'long' ? price + r * (cfg.rr_target ?? 1.5) : price - r * (cfg.rr_target ?? 1.5);

  return {
    signal: {
      strategy: 'liquidation_fade',
      direction,
      entry: price,
      stop, target,
      expectedHoldingMinutes: cfg.max_holding_minutes ?? 8,
      preferredEntryMode: 'ioc',
      trailType: 'none',
      reason: `LIQ_FADE ${direction.toUpperCase()} | longLiq=$${cascade.longUsd.toFixed(0)} shortLiq=$${cascade.shortUsd.toFixed(0)}`,
      features: {
        long_liq_usd: cascade.longUsd, short_liq_usd: cascade.shortUsd,
        atr: a1, swept_level: direction === 'long' ? sweep.prevLow : sweep.prevHigh,
      },
    },
  };
}
